### PR TITLE
Do dynamic layout of generic/resilient default actors properly

### PR DESCRIPTION
**Pull request by rjmccall**
Since these types have an implicit stored property, this requires adding an abstraction over fields to IRGen, at least throughout the class code.  In some ways I think this significantly improves the code, especially in how we approach missing members.

Fixes rdar://72202671.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test

File: lib/IRGen/Field.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/// directly accessing the stored properties is good practice.
+unsigned getNumFields(const NominalTypeDecl *type);
+
+struct Field {
+public:
+  enum Kind: uintptr_t {
+    Var,
+    MissingMember,
+    DefaultActorStorage,
+    FirstArtificial = DefaultActorStorage
```
Comment (varungandhi-apple) [inline lib/IRGen/Field.h:48]: Could you add a short description for what 'artificial' means in this context/why the actor storage can't be represented as an ordinary var?

File: lib/IRGen/Field.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  /// imported declarations.
+  bool isConcrete() const { return getKind() != MissingMember; }
+
+  /// Return the type of this concrete field.
+  SILType getType(IRGenModule &IGM, SILType baseType) const;
+
+  /// Return the interface type of this concrete field.
+  Type getInterfaceType(IRGenModule &IGM) const;
+
+  /// Return the nam eof this concrete field.
```
Comment (varungandhi-apple) [inline lib/IRGen/Field.h:91]: Typo: nam eof -> name of

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test Windows platform

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
LLDB failure that doesn't look related to my PR.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test OS X

**Pull request merged**